match_phrase query over match_only_text field can throw a NPE

When executing match_phrase over a match_only_text field, Elasticsearch use our own Lucene query implementation called `SourceConfirmedTextQuery`. It is easy to prove that running this query over an empty index with throw an NPE:

This test:
```
public void testEmptyIndex() throws Exception {
        try (Directory dir = newDirectory(); IndexWriter w = new IndexWriter(dir, newIndexWriterConfig(Lucene.STANDARD_ANALYZER))) {
            try (IndexReader reader = DirectoryReader.open(w)) {
                IndexSearcher searcher = new IndexSearcher(reader);
                PhraseQuery query = new PhraseQuery("body", "b", "c");
                Query sourceConfirmedPhraseQuery = new SourceConfirmedTextQuery(query, SOURCE_FETCHER_PROVIDER, Lucene.STANDARD_ANALYZER);
                searcher.search(sourceConfirmedPhraseQuery, 10);
            }
        }
    }
```

Produces the following exception:

```
annot invoke "org.apache.lucene.search.CollectionStatistics.sumTotalTermFreq()" because "collectionStats" is null
java.lang.NullPointerException: Cannot invoke "org.apache.lucene.search.CollectionStatistics.sumTotalTermFreq()" because "collectionStats" is null
	at __randomizedtesting.SeedInfo.seed([999E06EF36C5BA05:1D6D0A0AE74C9FFC]:0)
	at org.apache.lucene.search.similarities.BM25Similarity.avgFieldLength(BM25Similarity.java:110)
	at org.apache.lucene.search.similarities.BM25Similarity.scorer(BM25Similarity.java:203)
	at org.elasticsearch.index.mapper.extras.SourceConfirmedTextQuery.createWeight(SourceConfirmedTextQuery
```

This can happen in Elasticsearch too under some circumstances. I was able to reproduce the same behaviour in the following test:

```
 public void testSearch() throws IOException {

        prepareCreate("test")
            .setMapping(
                jsonBuilder().startObject()
                    .startObject("_doc")
                    .startObject("properties")
                    .startObject("foo").field("type", "match_only_text").endObject()
                    .endObject()
                    .endObject()
                    .endObject()
            )
            .get();
        ensureGreen();
        assertEquals(
            DocWriteResponse.Result.CREATED,
            client().prepareIndex("test").setSource("foo", "foo").get().getResult()
        );
        // don't refresh the index
        QueryBuilder queryBuilder1 = new MatchPhrasePrefixQueryBuilder("foo", "some complex text");
        SearchRequestBuilder request = client().prepareSearch("test");
        request.setQuery(
            QueryBuilders.boolQuery().filter(QueryBuilders.boolQuery().should(queryBuilder1).minimumShouldMatch(1)));
        request.get();
    }
```

With the error:

```
Caused by: java.lang.NullPointerException: Cannot invoke "org.apache.lucene.search.CollectionStatistics.field()" because "collectionStats" is null
	at org.apache.lucene.search.similarities.PerFieldSimilarityWrapper.scorer(PerFieldSimilarityWrapper.java:44)
	at org.elasticsearch.index.mapper.extras.SourceConfirmedTextQuery.createWeight(SourceConfirmedTextQuery.java:241)
	at org.apache.lucene.search.IndexSearcher.createWeight(IndexSearcher.java:871)
	at org.elasticsearch.search.internal.ContextIndexSearcher.createWeight(ContextIndexSearcher.java:161)
	at org.apache.lucene.search.ConstantScoreQuery.createWeight(ConstantScoreQuery.java:112)

```

--- a/modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/SourceConfirmedTextQuery.java
+++ b/modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/SourceConfirmedTextQuery.java
@@ -219,28 +219,35 @@ public final class SourceConfirmedTextQuery extends Query {
             throw new IllegalStateException("Query " + in + " doesn't have any term");
         }
         final String field = terms.iterator().next().field();
-        final Map<Term, TermStates> termStates = new HashMap<>();
-        final List<TermStatistics> termStats = new ArrayList<>();
-        for (Term term : terms) {
-            TermStates ts = termStates.computeIfAbsent(term, t -> {
-                try {
-                    return TermStates.build(searcher.getTopReaderContext(), t, scoreMode.needsScores());
-                } catch (IOException e) {
-                    throw new UncheckedIOException(e);
+        final CollectionStatistics collectionStatistics = searcher.collectionStatistics(field);
+        final SimScorer simScorer;
+        final Weight approximationWeight;
+        if (collectionStatistics == null) {
+            // field does not exist in the index
+            simScorer = null;
+            approximationWeight = null;
+        } else {
+            final Map<Term, TermStates> termStates = new HashMap<>();
+            final List<TermStatistics> termStats = new ArrayList<>();
+            for (Term term : terms) {
+                TermStates ts = termStates.computeIfAbsent(term, t -> {
+                    try {
+                        return TermStates.build(searcher.getTopReaderContext(), t, scoreMode.needsScores());
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
+                    }
+                });
+                if (scoreMode.needsScores()) {
+                    if (ts.docFreq() > 0) {
+                        termStats.add(searcher.termStatistics(term, ts.docFreq(), ts.totalTermFreq()));
+                    }
+                } else {
+                    termStats.add(new TermStatistics(term.bytes(), 1, 1L));
                 }
-            });
-            if (scoreMode.needsScores()) {
-                if (ts.docFreq() > 0) {
-                    termStats.add(searcher.termStatistics(term, ts.docFreq(), ts.totalTermFreq()));
-                }
-            } else {
-                termStats.add(new TermStatistics(term.bytes(), 1, 1L));
             }
+            simScorer = searcher.getSimilarity().scorer(boost, collectionStatistics, termStats.toArray(TermStatistics[]::new));
+            approximationWeight = searcher.createWeight(approximate(in), ScoreMode.COMPLETE_NO_SCORES, 1f);
         }
-        final SimScorer simScorer = searcher.getSimilarity()
-            .scorer(boost, searcher.collectionStatistics(field), termStats.toArray(TermStatistics[]::new));
-        final Weight approximationWeight = searcher.createWeight(approximate(in), ScoreMode.COMPLETE_NO_SCORES, 1f);
-
         return new Weight(this) {
 
             @Override
@@ -272,7 +279,7 @@ public final class SourceConfirmedTextQuery extends Query {
 
             @Override
             public RuntimePhraseScorer scorer(LeafReaderContext context) throws IOException {
-                final Scorer approximationScorer = approximationWeight.scorer(context);
+                final Scorer approximationScorer = approximationWeight != null ? approximationWeight.scorer(context) : null;
                 if (approximationScorer == null) {
                     return null;
                 }

--- a/modules/mapper-extras/src/test/java/org/elasticsearch/index/mapper/extras/SourceConfirmedTextQueryTests.java
+++ b/modules/mapper-extras/src/test/java/org/elasticsearch/index/mapper/extras/SourceConfirmedTextQueryTests.java
@@ -417,4 +417,16 @@ public class SourceConfirmedTextQueryTests extends ESTestCase {
         ).build();
         assertEquals(approximation, SourceConfirmedTextQuery.approximate(phrasePrefixQuery));
     }
+
+    public void testEmptyIndex() throws Exception {
+        try (Directory dir = newDirectory(); IndexWriter w = new IndexWriter(dir, newIndexWriterConfig(Lucene.STANDARD_ANALYZER))) {
+            try (IndexReader reader = DirectoryReader.open(w)) {
+                IndexSearcher searcher = new IndexSearcher(reader);
+                PhraseQuery query = new PhraseQuery("body", "a", "b");
+                Query sourceConfirmedPhraseQuery = new SourceConfirmedTextQuery(query, SOURCE_FETCHER_PROVIDER, Lucene.STANDARD_ANALYZER);
+                assertEquals(0, searcher.count(sourceConfirmedPhraseQuery));
+            }
+        }
+    }
+
 }

